### PR TITLE
Use esl/fast_scram to compute salted_password

### PR DESCRIPTION
**Pull request by michalwski**
This PR aims to improve the performance of computing scram password hashes.
It uses NIF-based implementaion of the `hi` function used to compute `salted_password`.

### CODE REVIEW

File: src/mongooseim.app.src [hunk 1]
```diff
@@ -14,6 +14,7 @@
                   cowboy,
                   cpool,
                   crypto,
+                  fast_scram,
```
Comment (NelsonVides) [inline src/mongooseim.app.src:17]: It is a stateless library, so I don't know if it is required to request otp to start it here. Is there some convention about the `applications` section maybe I'm missing? 🙂

**Comment by mongoose-im** [comment]
[8358.1](https://travis-ci.org/esl/MongooseIM/jobs/706521607) / Erlang 22.0 / small_tests / 6184ed11d3ebcedf7112fd62ec848efd35a40c84
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2787/8358/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2787/8358/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-07-09_13.26.50/lib.mongooseim.logs/run.2020-07-09_13.26.51/suite.log.html)

File: src/mongoose_scram.erl [hunk 2]
```diff
@@ -85,18 +85,6 @@ client_proof_key(ClientProof, ClientSignature) ->
...
-
--spec hi_round(sha_type(), binary(), binary(), non_neg_integer()) -> binary().
-hi_round(Sha, Password, UPrev, 1) ->
-    crypto:hmac(Sha, Password, UPrev);
-hi_round(Sha, Password, UPrev, IterationCount) ->
-    U = crypto:hmac(Sha, Password, UPrev),
-    mask(U, hi_round(Sha, Password, U, IterationCount - 1)).
-
 -spec mask(binary(), binary()) -> binary().
 mask(Key, Data) ->
```
Comment (NelsonVides) [inline src/mongoose_scram.erl:89]: By the way, I believe this `mask` function is equivalent to `crypto:exor/2`, so maybe we can also just use standard library functions and get rid of this one as well, it is used now in only one place. It even seems to have a very slightly better performance locally for me.

File: src/mongooseim.app.src (hunk 1 above)
Comment (michalwski) [reply src/mongooseim.app.src:17]: There is no need to start the app, on the other hand, the app needs to be included in the release. I think this is the clearest way.

**Comment by michalwski** [comment]
The load test is currently in pending state here: https://tide.erlang-solutions.com/public/fad2513e-e1d7-466c-9018-84e794923513

**Comment by mongoose-im** [comment]
[8359.1](https://travis-ci.org/esl/MongooseIM/jobs/706547217) / Erlang 22.0 / small_tests / 60ff327916bdc8639a1d531f64cd61a786791058
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2787/8359/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2787/8359/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-07-09_14.36.30/lib.mongooseim.logs/run.2020-07-09_14.36.31/suite.log.html)

**Comment by NelsonVides** [comment]
> The load test is currently in pending state here: https://tide.erlang-solutions.com/public/fad2513e-e1d7-466c-9018-84e794923513

Manual checking seems to say that the starting phase in Tide, where user registration takes place, took 13min, against >60min that was apparently taking before.

**Comment by Neustradamus** [comment]
Thanks a lot to all for your job about SCRAM!

**Pull request merged**